### PR TITLE
sql/driver: send DInterval over the wire as int64

### LINKED ISSUE #2397: sql: DInterval should come over the wire as something other than a string

### PR DESCRIPTION
**Pull request by tamird**
Fixes #2397.

The changes in logic test are unfortunate, but I think it's the best we can do with Go's `sql` abstraction.

### CODE REVIEW

**Comment by tamird** [comment]
Hm, added a commit that passes `time.Duration` as a `sql/driver.Value`, and it seems to work.

Any thoughts on if that's sane to do?

File: sql/driver/wire.proto [hunk 1]
```diff
@@ -48,6 +48,7 @@ message Datum {
     bytes bytes_val = 4;
     string string_val = 5;
     Timestamp time_val = 6;
+    int64 interval_val = 7;
```
Comment (vivekmenezes) [inline sql/driver/wire.proto]: cool. I suppose we can also add a
int64 date_val in the future, or in this change if you are so inclined.

Comment (tamird) [inline sql/driver/wire.proto]: sure, done.

File: sql/driver/wire.proto [hunk 2]
```diff
@@ -47,10 +47,9 @@ message Datum {
     double float_val = 3;
     bytes bytes_val = 4;
     string string_val = 5;
-    Timestamp time_val = 6;
-    // TODO(vivek): Add additional types like Date and Interval that are
-    // supported by Cockroach but not supported by the Go sql driver. These
-    // will be useful for drivers in other languages.
+    Timestamp date_val = 6;
```
Comment (vivekmenezes) [inline sql/driver/wire.proto]: We could also choose to use a int64 here since a date doesn't need nanoseconds. But I'm also happy leaving it in as a timestamp type to make the code more readable.

File: sql/executor.go [hunk 3]
```diff
@@ -285,6 +286,8 @@ func (p parameters) Arg(name string) (parser.Datum, bool) {
 		return parser.DString(t.StringVal), true
 	case *driver.Datum_TimeVal:
 		return parser.DTimestamp{Time: t.TimeVal.GoTime().UTC()}, true
+	case *driver.Datum_IntervalVal:
```
Comment (vivekmenezes) [inline sql/executor.go]: and Datum_DateVal?

**Comment by vivekmenezes** [comment]
LGTM

File: sql/executor.go (hunk 3 above)
Comment (tamird) [inline sql/executor.go]: good point, added. it's a shame we can't exercise this since the Go driver will never send `Datum_DateVal`.

File: sql/driver/wire.proto (hunk 2 above)
Comment (tamird) [inline sql/driver/wire.proto]: Agreed, but for now this is simpler

File: sql/executor.go [hunk 4]
```diff
@@ -283,8 +284,12 @@ func (p parameters) Arg(name string) (parser.Datum, bool) {
 		return parser.DBytes(t.BytesVal), true
 	case *driver.Datum_StringVal:
 		return parser.DString(t.StringVal), true
+	case *driver.Datum_DateVal:
+		return parser.DTimestamp{Time: t.DateVal.GoTime().UTC()}, true
```
Comment (petermattis) [inline sql/executor.go]: Per Ben's comment in another change, should this be `GoUTCTime()` or something like that?

**Comment by petermattis** [comment]
Using `time.Duration` for `driver.Value` seems to violate the `driver.Value` docs. But the docs seem much more restrictive than the actual implementation. [convertAssign](https://github.com/golang/go/blob/master/src/database/sql/convert.go#L85) is used to convert and assign a value returned by the driver to a pointer value passed to `Scan()`. The routine seems to have all sorts of logic for conversions and will happily convert a `time.Duration` to an `int64` because they have the same underlying `kind`.

File: sql/executor.go (hunk 4 above)
Comment (tamird) [inline sql/executor.go]: Sure. Done.

**Comment by tamird** [comment]
@petermattis I'll take that as an LGTM, and go ahead and keep https://github.com/tamird/cockroach/commit/f382546e3da8131e284ada89d2b895aee825decc.

Will merge on green.

File: sql/driver/wire.go [hunk 5]
```diff
@@ -80,14 +80,15 @@ func (d Datum) Value() (driver.Value, error) {
...
-		val = t.TimeVal.GoTime().UTC()
+		val = t.TimeVal.GoUTCTime()
+	case *Datum_IntervalVal:
+		val = time.Duration(t.IntervalVal)
 	}
 
-	if driver.IsValue(val) {
-		return val, nil
-	}
-	return nil, util.Errorf("unsupported type %T", val)
```
Comment (bdarnell) [inline sql/driver/wire.go]: Why did you remove this error?

**Comment by bdarnell** [comment]
LGTM

File: sql/driver/wire.go (hunk 5 above)
Comment (tamird) [inline sql/driver/wire.go]: Because time.Duration is indeed an unsupported type. See the commit message: we are deliberately violating the `driver.Value` contract here, but it seems to work.

Comment (bdarnell) [inline sql/driver/wire.go]: Ah, I see. It's the `else` for the `IsValue` check; I was reading it as the `default` for the type switch. Speaking of which, we should probably have one.

File: sql/executor.go (hunk 4 above)
Comment (tamird) [inline sql/executor.go]: Actually, that's not the right thing to do. We shouldn't be blindly calling `UTC()` on everything; `UTC()` only changes the behaviour of `time.Time` for presentational purposes, not computational.

I've added a commit here to remove all the non-presentational uses of `UTC()`, PTAL.

File: sql/driver/wire.go (hunk 5 above)
Comment (tamird) [inline sql/driver/wire.go]: Oh, you mean an error in the `default`? Sure.

File: sql/driver/wire.go [hunk 6]
```diff
@@ -80,14 +100,17 @@ func (d Datum) Value() (driver.Value, error) {
 		val = t.BytesVal
 	case *Datum_StringVal:
 		val = t.StringVal
+	case *Datum_DateVal:
+		val = Date{t.DateVal.GoTime()}
```
Comment (bdarnell) [inline sql/driver/wire.go]: +1 on the introduction of a distinct Date type, but this demonstrates why (at least for dates), the prior use of `.UTC()` was not just presentational. We represent dates as timestamps whose hour/minute/second components are all zero, but those components are only defined with respect to some timezone. At least for dates, we must associate a timezone with the `time.Time` instead of leaving it to be interpreted according to the local time (or we adjust the UTC date timestamp provided by the server for the local time zone)

I'm not sure what this means for timestamps. It doesn't seem quite right for the driver to set them to display in UTC, but it's also weird to have them inconsistent with dates. I'd probably leave them in UTC for now until we have a more coherent plan for how we're going to handle non-UTC timezones (including #2327). I suspect that the solution will involve server-side awareness of the client's timezone, so the driver will just be reflecting a timezone provided by the server instead of making these decisions on its own.

Comment (tamird) [inline sql/driver/wire.go]: Ah, I agree with your analysis regarding the timezone-awareness of Date. I think we can solve the Timestamp inconsistency by unexporting `Date.Time` and only exposing a constructor that `UTC()`s incoming `time.Time`s; I think that would suffice for removing any perceived symmetry between the types. WDYT?

Comment (bdarnell) [inline sql/driver/wire.go]: The inconsistency I was referring to was that if someone does `select now()::date, now()`, they may get something like ('2015-09-21', '2015-09-20 20:00:00') depending on their timezone. It's going to be confusing that timestamps display in local time but all other date math is done in UTC. I think that returning everything in UTC is a better short-term solution until we can make the server-side date math aware of the user's desired timezone.

Comment (tamird) [inline sql/driver/wire.go]: AFAIK, we never print timestamps with the time zone stripped. Any user that `Scan`s a timestamp into a `time.Time` and prints it with the time zone stripped is asking for trouble.

Comment (bdarnell) [inline sql/driver/wire.go]: Even with the time zone printed, it's surprising, is all I'm saying. It seems more consistent to let the display reflect the fact that the server does everything in UTC. But hopefully this all becomes moot soon enough when we teach the server about time zones.

Comment (tamird) [inline sql/driver/wire.go]: Reading this again, won't `select now()::date, now()` also print the date in the local time zone? I think the result will be ('2015-09-20', '2015-09-20 20:00:00').

Comment (bdarnell) [inline sql/driver/wire.go]: > I think the result will be ('2015-09-20', '2015-09-20 20:00:00').

Not if we make the proposed change to force UTC for dates (and if we don't make that change then we've just moved which hours of the day will have incorrect results).

> in other words, I think the current implementation is correct, and that someInterval::date is a rounding operation which always strips hour, minute, second.

Not as long as the server and client are working in different timezones. The server sets hours, minutes and seconds to zero, then when the client tries to print a local date it applies the timezone adjustment and discards hours again.

Comment (tamird) [inline sql/driver/wire.go]: > The server sets hours, minutes and seconds to zero

Where does the server do this?

Comment (bdarnell) [inline sql/driver/wire.go]: https://github.com/cockroachdb/cockroach/blob/af72b76c1dff094cfe667f8b1b124081f7ba6452/sql/parser/eval.go#L1067

If the server didn't truncate dates then `==` and other operators for dates would be broken.

Comment (tamird) [inline sql/driver/wire.go]: Ah, but this truncation is zone independent[0], it always truncates relative to the zero time. We do probably want to make that truncation client time zone aware, but this PR doesn't make the behaviour any worse. I guess the current behaviour seems to be consistent because `Truncate` happens to operate in UTC, but it's just a coincidence. 

[0] http://play.golang.org/p/HAd47eseDg

Comment (bdarnell) [inline sql/driver/wire.go]: The truncation in `Time.Truncate` is zone independent (i.e. always UTC), but the "truncation" implied by d.Format("2006-01-02") is not. http://play.golang.org/p/bzvgCPfD8a

Comment (tamird) [inline sql/driver/wire.go]: I added a commit that replaces `Time.Truncate` with a zone-aware implementation - I'm pretty certain at this point that Timestamp and Date are consistent, but could you have another look?

Comment (bdarnell) [inline sql/driver/wire.go]: The client and server each have a time zone to be aware of; this only works when they are the same. I believe we will need #2327 to have a good time zone policy; until then the only consistent policy is to use UTC everywhere. Introducing new dependencies on the server's system timezone is a step in the wrong direction.

Comment (tamird) [inline sql/driver/wire.go]: OK, I've backed out the last commit. Will revisit when #2327 is implemented.
